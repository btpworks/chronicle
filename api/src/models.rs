use crate::schema::*;
use diesel::{Insertable, Queryable};

#[derive(Queryable)]
pub struct NameSpace {
    pub name: String,
    pub uuid: String,
}

#[derive(Insertable)]
#[table_name = "namespace"]
pub struct NewNamespace<'a> {
    pub name: &'a str,
    pub uuid: &'a str,
}

#[derive(Queryable)]
pub struct Agent {
    pub name: String,
    pub namespace: String,
    pub publickey: Option<String>,
    pub privatekeypath: Option<String>,
    pub current: i32,
}

#[derive(Insertable)]
#[table_name = "agent"]
pub struct NewAgent<'a> {
    pub name: &'a str,
    pub namespace: &'a str,
    pub current: i32,
    pub publickey: Option<&'a str>,
    pub privatekeypath: Option<&'a str>,
}
