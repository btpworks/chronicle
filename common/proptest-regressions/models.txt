# Seeds for failure cases proptest has generated in the past. It is
# automatically read and these particular cases re-run before any
# novel cases are generated.
#
# It is recommended to check this file in to source control so that
# everyone who runs the test benefits from these saved cases.
cc 33eb56f784221d700a7a878e07454fe00e91f4573e017e4b3bf3f2dee5f2326c # shrinks to tx = [StartActivity(StartActivity { namespace: NamespaceId("chronicle:ns:a:00000000-0000-0000-0000-000000000000"), id: ActivityId("chronicle:activity:A"), agent: AgentId("chronicle:agent:A"), time: 2021-11-29T00:00:00Z })]
