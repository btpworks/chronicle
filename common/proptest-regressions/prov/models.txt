# Seeds for failure cases proptest has generated in the past. It is
# automatically read and these particular cases re-run before any
# novel cases are generated.
#
# It is recommended to check this file in to source control so that
# everyone who runs the test benefits from these saved cases.
cc 5d903914c5be8a19427ef50530a7588d4dff9e7b6ab246f86c06ca41708da4f9 # shrinks to tx = [ActivityUses(ActivityUses { namespace: NamespaceId("http://blockchaintp.com/chronicle/ns#ns:ik0Yowd+Nh-37-uY8ECN2-b-FCX+960:00000000-0000-0000-0000-000000000000"), id: EntityId("http://blockchaintp.com/chronicle/ns#entity:a"), activity: ActivityId("http://blockchaintp.com/chronicle/ns#activity:a") })]
