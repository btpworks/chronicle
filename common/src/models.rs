use async_std::task::block_on;
use chrono::{DateTime, Utc};
use iref::{AsIri, Iri};
use json::{object, JsonValue};
use json_ld::{context::Local, Document, JsonContext, NoLoader};
use multimap::MultiMap;
use std::collections::HashMap;
use uuid::Uuid;

use crate::vocab::{Chronicle, Prov};

#[derive(Serialize, Deserialize, PartialEq, Eq, Hash, Debug, Clone)]
pub struct NamespaceId(String);

impl std::ops::Deref for NamespaceId {
    type Target = String;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl NamespaceId {
    /// Decompose a namespace id into its constituent parts, we need to preserve the type better to justify this implementation
    pub fn decompose(&self) -> (&str, Uuid) {
        if let &[_, _, _, name, uuid, ..] = &self.0.split(":").collect::<Vec<_>>()[..] {
            return (name, Uuid::parse_str(uuid).unwrap());
        }

        unreachable!();
    }
}

impl<S> From<S> for NamespaceId
where
    S: AsIri,
{
    fn from(iri: S) -> Self {
        Self(iri.as_iri().to_string())
    }
}

#[derive(Serialize, Deserialize, PartialEq, Eq, Hash, Debug, Clone)]
pub struct EntityId(String);

impl std::ops::Deref for EntityId {
    type Target = String;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<S> From<S> for EntityId
where
    S: AsIri,
{
    fn from(iri: S) -> Self {
        Self(iri.as_iri().to_string())
    }
}

#[derive(Serialize, Deserialize, PartialEq, Eq, Hash, Debug, Clone)]
pub struct AgentId(String);

impl std::ops::Deref for AgentId {
    type Target = String;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl AgentId {
    /// Extract the agent name from an id
    pub fn decompose(&self) -> &str {
        if let &[_, _, _, name, ..] = &self.0.split(":").collect::<Vec<_>>()[..] {
            return name;
        }

        unreachable!();
    }
}

impl<S> From<S> for AgentId
where
    S: AsIri,
{
    fn from(iri: S) -> Self {
        Self(iri.as_iri().to_string())
    }
}

#[derive(Serialize, Deserialize, PartialEq, Eq, Hash, Debug, Clone)]
pub struct ActivityId(String);

impl std::ops::Deref for ActivityId {
    type Target = String;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl ActivityId {
    /// Extract the activity name from an id
    pub fn decompose(&self) -> &str {
        if let &[_, _, _, name, ..] = &self.0.split(":").collect::<Vec<_>>()[..] {
            return name;
        }

        unreachable!();
    }
}

impl<S> From<S> for ActivityId
where
    S: AsIri,
{
    fn from(iri: S) -> Self {
        Self(iri.as_iri().to_string())
    }
}

#[derive(Serialize, Deserialize, PartialEq, Debug, Clone)]
pub struct CreateNamespace {
    pub id: NamespaceId,
    pub name: String,
    pub uuid: Uuid,
}

#[derive(Serialize, Deserialize, PartialEq, Debug, Clone)]
pub struct CreateAgent {
    pub namespace: NamespaceId,
    pub name: String,
    pub id: AgentId,
}

#[derive(Serialize, Deserialize, PartialEq, Debug, Clone)]
pub struct RegisterKey {
    pub namespace: NamespaceId,
    pub id: AgentId,
    pub publickey: String,
    pub name: String,
}

#[derive(Serialize, Deserialize, PartialEq, Debug, Clone)]
pub struct CreateActivity {
    pub namespace: NamespaceId,
    pub id: ActivityId,
    pub name: String,
}

#[derive(Serialize, Deserialize, PartialEq, Debug, Clone)]
pub struct StartActivity {
    pub namespace: NamespaceId,
    pub id: ActivityId,
    pub agent: AgentId,
    pub time: DateTime<Utc>,
}

#[derive(Serialize, Deserialize, PartialEq, Debug, Clone)]
pub struct ActivityUses {
    pub namespace: NamespaceId,
    pub id: ActivityId,
    pub entity: EntityId,
}

#[derive(Serialize, Deserialize, PartialEq, Debug, Clone)]
pub struct GenerateEntity {
    pub namespace: NamespaceId,
    pub id: EntityId,
    pub activity: ActivityId,
}

#[derive(Serialize, Deserialize, PartialEq, Debug, Clone)]
pub enum ChronicleTransaction {
    CreateNamespace(CreateNamespace),
    CreateAgent(CreateAgent),
    RegisterKey(RegisterKey),
    CreateActivity(CreateActivity),
    StartActivity(StartActivity),
    ActivityUses(ActivityUses),
    GenerateEntity(GenerateEntity),
}

#[derive(Debug, Clone)]
pub struct Namespace {
    pub id: NamespaceId,
    pub uuid: Uuid,
    pub name: String,
}

impl Namespace {
    pub fn new(id: NamespaceId, uuid: Uuid, name: String) -> Self {
        Self { id, uuid, name }
    }
}

#[derive(Debug, Clone)]
pub struct Agent {
    pub id: AgentId,
    pub namespaceid: NamespaceId,
    pub name: String,
    pub publickey: Option<String>,
}

impl Agent {
    pub fn new(
        id: AgentId,
        namespaceid: NamespaceId,
        name: String,
        publickey: Option<String>,
    ) -> Self {
        Self {
            id,
            namespaceid,
            name,
            publickey,
        }
    }
}

#[derive(Debug, Clone)]
pub struct Activity {
    pub id: ActivityId,
    pub namespaceid: NamespaceId,
    pub name: String,
    pub started: Option<DateTime<Utc>>,
    pub ended: Option<DateTime<Utc>>,
}

impl Activity {
    pub fn new(id: ActivityId, ns: NamespaceId, name: &str) -> Self {
        Self {
            id,
            namespaceid: ns,
            name: name.to_owned(),
            started: None,
            ended: None,
        }
    }
}

#[derive(Debug, Clone)]
pub struct Entity {
    id: EntityId,
    ns: NamespaceId,
}

impl Entity {
    pub fn new(id: EntityId, ns: NamespaceId) -> Self {
        Self { id, ns }
    }
}

#[derive(Debug, Default)]
pub struct ProvModel {
    pub namespaces: HashMap<NamespaceId, Namespace>,
    pub agents: HashMap<AgentId, Agent>,
    pub activities: HashMap<ActivityId, Activity>,
    pub entities: HashMap<EntityId, Entity>,
    pub was_associated_with: MultiMap<ActivityId, AgentId>,
    pub was_attributed_to: MultiMap<EntityId, AgentId>,
    pub was_generated_by: MultiMap<EntityId, ActivityId>,
    pub uses: MultiMap<ActivityId, EntityId>,
}

impl ProvModel {
    pub fn from_tx<'a, I>(tx: I) -> Self
    where
        I: IntoIterator<Item = &'a ChronicleTransaction>,
    {
        let mut model = Self::default();
        for tx in tx {
            model.apply(tx);
        }

        model
    }

    fn namespace_context(&mut self, ns: &NamespaceId) {
        let (namespacename, uuid) = ns.decompose();

        self.namespaces.insert(
            ns.clone(),
            Namespace {
                id: ns.clone(),
                uuid: uuid,
                name: namespacename.to_owned(),
            },
        );
    }

    /// Transform a sequence of ChronicleTransaction events into a provenance model,
    /// If a statement requires a subject or object that does not currently exist in the model, then we create it
    pub fn apply(&mut self, tx: &ChronicleTransaction) {
        let tx = tx.to_owned();
        match tx {
            ChronicleTransaction::CreateNamespace(CreateNamespace { id, name, uuid }) => {
                self.namespaces
                    .insert(id.clone(), Namespace::new(id, uuid, name).into());
            }
            ChronicleTransaction::CreateAgent(CreateAgent {
                namespace,
                id,
                name,
            }) => {
                self.namespace_context(&namespace);
                self.agents
                    .insert(id.clone(), Agent::new(id, namespace, name, None));
            }
            ChronicleTransaction::RegisterKey(RegisterKey {
                namespace,
                id,
                publickey,
                name,
            }) => {
                self.namespace_context(&namespace);

                if !self.agents.contains_key(&id) {
                    self.agents
                        .insert(id.clone(), Agent::new(id.clone(), namespace, name, None));
                }
                self.agents
                    .get_mut(&id)
                    .map(|x| x.publickey = Some(publickey));
            }
            ChronicleTransaction::CreateActivity(CreateActivity {
                namespace,
                id,
                name,
            }) => {
                self.namespace_context(&namespace);

                if !self.activities.contains_key(&id) {
                    self.activities
                        .insert(id.clone(), Activity::new(id, namespace, &name));
                }
            }
            ChronicleTransaction::StartActivity(StartActivity {
                namespace,
                id,
                agent,
                time,
            }) => {
                self.namespace_context(&namespace);
                if !self.activities.contains_key(&id) {
                    let activity_name = id.decompose();
                    let mut activity = Activity::new(id.clone(), namespace.clone(), activity_name);
                    activity.started = Some(time);
                    self.activities.insert(id.clone(), activity);
                }

                if !self.agents.contains_key(&agent) {
                    let agent_name = agent.decompose();
                    let agentmodel =
                        Agent::new(agent.clone(), namespace, agent_name.to_owned(), None);
                    self.agents.insert(agent.clone(), agentmodel);
                }

                self.was_associated_with
                    .entry(id)
                    .or_insert_vec(vec![])
                    .push(agent.clone())
            }
            ChronicleTransaction::ActivityUses(ActivityUses {
                namespace,
                id,
                entity,
            }) => {
                self.namespace_context(&namespace);
                if !self.activities.contains_key(&id) {
                    let activity_name = id.decompose();
                    self.activities.insert(
                        id.clone(),
                        Activity::new(id.clone(), namespace.clone(), activity_name),
                    );
                }
                if !self.entities.contains_key(&entity) {
                    self.entities
                        .insert(entity.clone(), Entity::new(entity.clone(), namespace));
                }

                self.uses.insert(id, entity);
            }
            ChronicleTransaction::GenerateEntity(GenerateEntity {
                namespace,
                id,
                activity,
            }) => {
                if !self.activities.contains_key(&activity) {
                    let activity_name = activity.decompose();
                    self.activities.insert(
                        activity.clone(),
                        Activity::new(activity.clone(), namespace.clone(), activity_name),
                    );
                }
                if !self.entities.contains_key(&id) {
                    self.entities
                        .insert(id.clone(), Entity::new(id.clone(), namespace));
                }

                self.was_generated_by.insert(id, activity);
            }
        };
    }

    /// Write the model out as a JSON-LD document in expanded form
    pub fn to_json(&self) -> ExpandedJson {
        let mut doc = json::Array::new();

        for (id, ns) in self.namespaces.iter() {
            doc.push(object! {
                "@id": (*id.as_str()),
                "@type": Iri::from(Chronicle::NamespaceType).as_str(),
                "http://www.w3.org/2000/01/rdf-schema#label": [{
                    "@value": ns.name.as_str(),
                }]
            })
        }

        for (id, agent) in self.agents.iter() {
            let mut agentdoc = object! {
                "@id": (*id.as_str()),
                "@type": Iri::from(Prov::Agent).as_str(),
                "http://www.w3.org/2000/01/rdf-schema#label": [{
                   "@value": agent.name.as_str(),
                }]
            };
            agent.publickey.as_ref().map(|publickey| {
                let mut values = json::Array::new();

                values.push(object! {
                    "@value": JsonValue::String(publickey.to_owned()),
                });

                agentdoc
                    .insert(Iri::from(Chronicle::HasPublicKey).as_str(), values)
                    .ok();
            });

            let mut values = json::Array::new();

            values.push(object! {
                "@id": JsonValue::String(agent.namespaceid.0.clone()),
            });

            agentdoc
                .insert(Iri::from(Chronicle::HasNamespace).as_str(), values)
                .ok();

            doc.push(agentdoc);
        }

        for (id, activity) in self.activities.iter() {
            let mut activitydoc = object! {
                "@id": (*id.as_str()),
                "@type": Iri::from(Prov::Activity).as_str(),
                "http://www.w3.org/2000/01/rdf-schema#label": [{
                   "@value": activity.name.as_str(),
                }]
            };

            activity.started.map(|time| {
                let mut values = json::Array::new();
                values.push(object! {"@value": time.to_rfc3339()});

                activitydoc
                    .insert("http://www.w3.org/ns/prov#startedAtTime", values)
                    .ok();
            });

            self.was_associated_with.get_vec(&id).map(|asoc| {
                let mut ids = json::Array::new();

                for id in asoc.iter() {
                    ids.push(object! {"@id": id.as_str()});
                }

                activitydoc
                    .insert("http://www.w3.org/ns/prov#wasAssociatedWith", ids)
                    .ok();
            });

            let mut values = json::Array::new();

            values.push(object! {
                "@id": JsonValue::String(activity.namespaceid.0.clone()),
            });

            activitydoc
                .insert(Iri::from(Chronicle::HasNamespace).as_str(), values)
                .ok();

            doc.push(activitydoc);
        }

        ExpandedJson(doc.into())
    }
}

pub struct ExpandedJson(pub JsonValue);

impl ExpandedJson {
    pub fn compact(self) -> Result<CompactedJson, json_ld::Error> {
        let processed_context =
            block_on(crate::context::PROV.process::<JsonContext, _>(&mut NoLoader, None))?;

        // Compaction.
        let output = block_on(self.0.compact(&processed_context, &mut NoLoader))?;

        Ok(CompactedJson(output))
    }
}

pub struct CompactedJson(pub JsonValue);

impl std::ops::Deref for CompactedJson {
    type Target = JsonValue;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
