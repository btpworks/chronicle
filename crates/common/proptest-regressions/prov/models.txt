# Seeds for failure cases proptest has generated in the past. It is
# automatically read and these particular cases re-run before any
# novel cases are generated.
#
# It is recommended to check this file in to source control so that
# everyone who runs the test benefits from these saved cases.
cc 5d903914c5be8a19427ef50530a7588d4dff9e7b6ab246f86c06ca41708da4f9 # shrinks to tx = [ActivityUses(ActivityUses { namespace: NamespaceId("chronicle:ns:ik0Yowd+Nh-37-uY8ECN2-b-FCX+960:00000000-0000-0000-0000-000000000000"), id: EntityId("chronicle:entity:a"), activity: ActivityId("chronicle:activity:a") })]
