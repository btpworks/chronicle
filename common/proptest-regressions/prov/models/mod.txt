# Seeds for failure cases proptest has generated in the past. It is
# automatically read and these particular cases re-run before any
# novel cases are generated.
#
# It is recommended to check this file in to source control so that
# everyone who runs the test benefits from these saved cases.
cc bd764dc5055ffe1c5844769b18625a40e9d1f3ec140af86c2a33d742f0e09f6c # shrinks to tx = [SetAttributes(Entity { namespace: NamespaceId("chronicle:ns:a:00000000-0000-0000-0000-000000000000"), id: EntityId("chronicle:entity:a"), attributes: Attributes { typ: Some(DomaintypeId("chronicle:domaintype:mSB8++e0u-Ja-o-lA-oo+K3g+WlwIu")), attributes: {"A": Attribute { typ: "A", value: String("data") }, "a": Attribute { typ: "a", value: String("data") }, "-": Attribute { typ: "-", value: String("data") }, "+": Attribute { typ: "+", value: String("data") }} } })]
