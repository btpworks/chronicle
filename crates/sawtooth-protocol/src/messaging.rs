use std::{collections::HashSet, sync::Arc};

use crate::{
    address::{SawtoothAddress, FAMILY, VERSION},
    messages::MessageBuilder,
    sawtooth::{Batch, ClientBatchSubmitRequest, ClientBatchSubmitResponse, Transaction},
};

use common::{
    k256::ecdsa::SigningKey,
    ledger::{LedgerWriter, SubmissionError},
    protocol::ProtocolError,
    prov::{ChronicleTransaction, ChronicleTransactionId, ProcessorError},
};
use custom_error::*;
use derivative::Derivative;
use prost::Message as ProstMessage;

use sawtooth_sdk::{
    messages::validator::Message_MessageType,
    messaging::{
        stream::{MessageConnection, MessageReceiver, MessageSender, ReceiveError, SendError},
        zmq_stream::{ZmqMessageConnection, ZmqMessageSender},
    },
};
use tokio::task::JoinError;
use tracing::{debug, instrument, trace};

#[derive(Derivative)]
#[derivative(Debug)]
pub struct SawtoothSubmitter {
    #[derivative(Debug = "ignore")]
    tx: ZmqMessageSender,
    rx: MessageReceiver,
    builder: OperationMessageBuilder,
    address: String,
}

#[derive(Debug)]
pub struct OperationMessageBuilder {
    builder: MessageBuilder,
}

impl OperationMessageBuilder {
    pub fn new(signer: &SigningKey, family: &str, version: &str) -> Self {
        OperationMessageBuilder {
            builder: MessageBuilder::new(signer.to_owned(), family, version),
        }
    }

    pub async fn make_tx(
        &mut self,
        transactions: &ChronicleTransaction,
    ) -> Result<(Transaction, ChronicleTransactionId), ProtocolError> {
        let addresses = transactions
            .tx
            .iter()
            .flat_map(|tx| tx.dependencies())
            .map(|addr| (SawtoothAddress::from(&addr).to_string(), addr))
            .collect::<HashSet<_>>();

        debug!(address_map = ?addresses);

        self.builder
            .make_sawtooth_transaction(
                addresses.iter().map(|x| x.0.clone()).collect(),
                addresses.into_iter().map(|x| x.0).collect(),
                vec![],
                transactions,
            )
            .await
    }

    pub fn wrap_tx_as_sawtooth_batch(&mut self, tx: Transaction) -> Batch {
        self.builder.wrap_tx_as_sawtooth_batch(tx)
    }

    pub async fn make_sawtooth_transaction(
        &mut self,
        inputs: Vec<String>,
        outputs: Vec<String>,
        dependencies: Vec<String>,
        payload: &ChronicleTransaction,
    ) -> Result<(Transaction, ChronicleTransactionId), ProtocolError> {
        self.builder
            .make_sawtooth_transaction(inputs, outputs, dependencies, payload)
            .await
    }
}

custom_error! {pub SawtoothSubmissionError
    Send{source: SendError}                                 = "Submission failed to send to validator",
    Recv{source: ReceiveError}                              = "Submission failed to send to validator",
    UnexpectedStatus{status: i32}                           = "Validator status unexpected {status}",
    Join{source: JoinError}                                 = "Submission blocking thread pool",
    Ld{source: ProcessorError}                              = "Json LD processing",
    Protocol{source: ProtocolError}                         = "Protocol {source}",
}

/// The sawtooth futures and their sockets are not controlled by a compatible reactor
impl SawtoothSubmitter {
    #[allow(dead_code)]
    pub fn new(address: &url::Url, signer: &SigningKey) -> Self {
        let builder = OperationMessageBuilder::new(signer, FAMILY, VERSION);
        let (tx, rx) = ZmqMessageConnection::new(address.as_str()).create();
        SawtoothSubmitter {
            tx,
            rx,
            builder,
            address: address.to_string(),
        }
    }

    #[instrument(
        name = "submit_sawtooth_tx",
        level = "info",
        skip(self, transactions),
        ret(Debug)
    )]
    async fn submit(
        &mut self,
        transactions: &ChronicleTransaction,
    ) -> Result<ChronicleTransactionId, (ChronicleTransactionId, SawtoothSubmissionError)> {
        // Practically, a protobuf serialization error here is probably a crash
        // loop level fault, but we will handle it without panic for now
        let (sawtooth_transaction, tx_id) = self
            .builder
            .make_tx(transactions)
            .await
            .map_err(|e| (ChronicleTransactionId::from(""), e.into()))?;

        let ret_tx_id = tx_id.clone();

        let res = async move {
            let batch = self.builder.wrap_tx_as_sawtooth_batch(sawtooth_transaction);

            trace!(?batch, "Validator request");

            let request = ClientBatchSubmitRequest {
                batches: vec![batch],
            };

            let mut future = loop {
                let future = self.tx.send(
                    Message_MessageType::CLIENT_BATCH_SUBMIT_REQUEST,
                    &tx_id.to_string(),
                    &request.encode_to_vec(),
                );

                // Force reconnection on any send error that's not a timeout -
                // disconnect can actually mean a dead Zmq Thread
                if let Err(SendError::UnknownError) | Err(SendError::DisconnectedError) = future {
                    debug!("Send error, re-initialise ZMQ");
                    let (tx, rx) = ZmqMessageConnection::new(self.address.as_str()).create();
                    self.tx = tx;
                    self.rx = rx;
                    continue;
                }

                break future.unwrap();
            };

            debug!(submit_transaction=%tx_id);

            let result = future.get_timeout(std::time::Duration::from_secs(10))?;

            let response =
                ClientBatchSubmitResponse::decode(&*result.content).map_err(ProtocolError::from)?;

            debug!(validator_response=?response);

            if response.status == 1 {
                Ok(tx_id)
            } else {
                Err(SawtoothSubmissionError::UnexpectedStatus {
                    status: response.status,
                })
            }
        }
        .await;

        res.map_err(|e| (ret_tx_id, e))
    }
}

#[async_trait::async_trait(?Send)]
impl LedgerWriter for SawtoothSubmitter {
    /// TODO: This blocks on a bunch of non tokio / futures 'futures' in the sawtooth rust SDK,
    /// which also exposes a bunch of non clonable types so we probably need another dispatch / join mpsc here
    async fn submit(
        &mut self,
        tx: &ChronicleTransaction,
    ) -> Result<ChronicleTransactionId, SubmissionError> {
        self.submit(tx)
            .await
            .map_err(|(tx_id, e)| SubmissionError::implementation(&tx_id, Arc::new(e.into())))
    }
}
