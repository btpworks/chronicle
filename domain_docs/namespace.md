# `chronicle:Namespace`

An IRI containing an external id and uuid part, used for disambiguation.
In order to work on the same namespace discrete Chronicle instances must share
the uuid part.
