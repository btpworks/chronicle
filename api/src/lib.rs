#[macro_use]
extern crate diesel;
#[macro_use]
extern crate diesel_migrations;
#[macro_use]
extern crate iref_enum;
#[macro_use]
extern crate json;

mod models;
mod schema;
mod vocab;

use async_std::task;
use custom_error::*;
use derivative::*;
use diesel::{prelude::*, sqlite::SqliteConnection};
use json::JsonValue;
use json_ld::{context::Local, Document, JsonContext, NoLoader};
use std::path::Path;

use common::{
    models::{ChronicleTransaction, CreateAgent, CreateNamespace},
    signing::{DirectoryStoredKeys, SignerError},
};
use iref::IriBuf;
use models::Agent;
use proto::messaging::{SawtoothValidator, SubmissionError};
use tracing::instrument;
use url::Url;
use user_error::UFE;
use uuid::Uuid;

embed_migrations!();

custom_error! {pub ApiError
    Db{source: diesel::result::Error}                           = "Database operation failed",
    DbConnection{source: diesel::ConnectionError}               = "Database connection failed",
    DbMigration{source: diesel_migrations::RunMigrationsError}  = "Database migration failed",
    Iri{source: iref::Error}                                    = "Invalid IRI",
    JsonLD{source: json_ld::Error}                              = "Json LD processing",
    Api{source: SubmissionError}                                = "Json LD processing",
    Signing{source: SignerError}                                = "Signing",
}

impl UFE for ApiError {}

#[derive(Debug)]
pub enum NamespaceCommand {
    Create { name: String },
}

#[derive(Debug)]
pub enum AgentCommand {
    Create {
        name: String,
        namespace: String,
    },
    RegisterKey {
        name: String,
        namespace: String,
        public: String,
        private: Option<String>,
    },
}

#[derive(Debug)]
pub enum ApiCommand {
    NameSpace(NamespaceCommand),
    Agent(AgentCommand),
}

#[derive(Debug)]
pub enum ApiResponse {
    Unit,
    Iri(IriBuf),
    Document(JsonValue),
}

#[derive(Derivative)]
#[derivative(Debug)]
pub struct Api {
    #[derivative(Debug = "ignore")]
    connection: SqliteConnection,
    ledger: SawtoothValidator,
}

impl Api {
    pub fn new(
        database_url: &str,
        sawtooth_url: &Url,
        secret_path: &Path,
    ) -> Result<Self, ApiError> {
        let connection = SqliteConnection::establish(database_url)?;
        embedded_migrations::run(&connection)?;

        let ledger = SawtoothValidator::new(
            sawtooth_url,
            DirectoryStoredKeys::new(secret_path)?.default(),
        );
        Ok(Api { connection, ledger })
    }

    #[instrument]
    fn create_namespace(&self, name: &str) -> Result<ApiResponse, ApiError> {
        let uuid = Uuid::new_v4();
        let newnamespace = models::NewNamespace {
            name,
            uuid: &uuid.to_string(),
        };

        diesel::insert_or_ignore_into(schema::namespace::table)
            .values(&newnamespace)
            .execute(&self.connection)
            .map(|_| {
                self.ledger
                    .submit(vec![ChronicleTransaction::CreateNamespace(
                        CreateNamespace {
                            id: IriBuf::from(&newnamespace).as_iri().into(),
                        },
                    )])
            })
            .map(|_| ApiResponse::Unit)
            .map_err(ApiError::from)
    }

    #[instrument]
    fn create_agent(&self, name: &str, namespace: &str) -> Result<ApiResponse, ApiError> {
        use self::schema::namespace::dsl as ns;

        self.create_namespace(namespace)?;

        let namespace_data = ns::namespace
            .filter(ns::name.eq(namespace))
            .first::<models::NameSpace>(&self.connection)?;

        let namespace_iri = IriBuf::new(&format!(
            "chronicle:{}/{}#",
            namespace_data.name, namespace_data.uuid
        ))?;

        let id = IriBuf::new(&format!("{}agent:{}", namespace_iri, name))?;

        let input = object! {
            "http://www.w3.org/ns/prov#Agent" : {
                "@id": (id.as_str())
            }
        };
        let context = object! {};

        let processed_context =
            task::block_on(context.process::<JsonContext, _>(&mut NoLoader, None))?;

        let output = task::block_on(input.compact(&processed_context, &mut NoLoader))?;

        let newagent = models::NewAgent {
            name,
            namespace,
            current: 0,
            publickey: None,
            privatekeypath: None,
        };

        diesel::insert_or_ignore_into(schema::agent::table)
            .values(&newagent)
            .execute(&self.connection)
            .map(|_| {
                self.ledger
                    .submit(vec![ChronicleTransaction::CreateAgent(CreateAgent {
                        id: IriBuf::from(&newagent).as_iri().into(),
                        namespace: IriBuf::from(&namespace_data).as_iri().into(),
                    })])
            })
            .map_err(ApiError::from)
            .map(|_| ApiResponse::Document(output))
    }

    #[instrument]
    pub fn dispatch(&self, command: ApiCommand) -> Result<ApiResponse, ApiError> {
        match command {
            ApiCommand::NameSpace(NamespaceCommand::Create { name }) => {
                self.create_namespace(&name)
            }
            ApiCommand::Agent(AgentCommand::Create { name, namespace }) => {
                self.create_agent(&name, &namespace)
            }
            ApiCommand::Agent(AgentCommand::RegisterKey {
                name,
                namespace,
                public,
                private,
            }) => self.register_key(&name, &namespace, &public, private),
        }
    }

    #[instrument]
    fn get_agent(
        &self,
        name: &str,
        namespace: &str,
    ) -> Result<Option<Agent>, diesel::result::Error> {
        use self::schema::agent::dsl as ns;
        ns::agent
            .filter(ns::name.eq(name).and(ns::namespace.eq(namespace)))
            .first::<models::Agent>(&self.connection)
            .optional()
    }

    #[instrument]
    fn register_key(
        &self,
        name: &str,
        namespace: &str,
        publickey: &str,
        privatekeypath: Option<String>,
    ) -> Result<ApiResponse, ApiError> {
        use self::schema::agent::dsl as ns;

        if let Some(Agent { current, .. }) = self.get_agent(name, namespace)? {
            let update = models::NewAgent {
                name,
                namespace,
                current,
                publickey: Some(publickey),
                privatekeypath: privatekeypath.as_deref(),
            };

            diesel::update(schema::agent::table)
                .filter(ns::name.eq(namespace).and(ns::namespace.eq(namespace)))
                .set(update)
                .execute(&self.connection)
                .map_err(ApiError::from)
                .map(|_| ApiResponse::Unit)
        } else {
            let insert = models::NewAgent {
                name,
                namespace,
                publickey: Some(publickey),
                privatekeypath: privatekeypath.as_deref(),
                ..Default::default()
            };

            diesel::insert_into(schema::agent::table)
                .values(insert)
                .execute(&self.connection)
                .map_err(ApiError::from)
                .map(|_| ApiResponse::Unit)
        }
    }
}
