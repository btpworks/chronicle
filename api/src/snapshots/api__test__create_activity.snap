---
source: api/src/lib.rs
expression: dump_ledger_state(api)

---
[
  {
    "namespace": "chronicle:ns:testns:5a0ab5b8-eeb7-4812-9fe3-6dd69bd20cea",
    "resource": "chronicle:activity:testactivity_0"
  },
  "{\"@id\":\"chronicle:activity:testactivity_0\",\"@type\":\"prov:Activity\",\"label\":\"testactivity_0\",\"namespace\":\"chronicle:ns:testns:5a0ab5b8-eeb7-4812-9fe3-6dd69bd20cea\"}",
  {
    "namespace": "null",
    "resource": "chronicle:ns:testns:5a0ab5b8-eeb7-4812-9fe3-6dd69bd20cea"
  },
  "{\"@id\":\"chronicle:ns:testns:5a0ab5b8-eeb7-4812-9fe3-6dd69bd20cea\",\"@type\":\"chronicle:Namespace\",\"label\":\"testns\"}"
]
